src/System.Private.CoreLib/shared/System/Threading/Tasks/FutureFactory.cs

```diff
@@ -1260,14 +1260,15 @@ internal static void CompleteFromAsyncResult(IAsyncResult asyncResult)
                 // Grab the relevant state and then null it out so that the task doesn't hold onto the state unnecessarily
                 var thisRef = promise!.m_thisRef; // TODO-NULLABLE: Remove ! when [DoesNotReturn] respected
                 var endMethod = promise.m_endMethod;
-                promise.m_thisRef = default!; // TODO-NULLABLE: Remove ! when nullable attributes are respected
+                promise.m_thisRef = default;
                 promise.m_endMethod = null;
                 if (endMethod == null) ThrowHelper.ThrowArgumentException(ExceptionResource.InvalidOperation_WrongAsyncResultOrEndCalledMultiple, ExceptionArgument.asyncResult);
 
                 // Complete the promise.  If the IAsyncResult completed synchronously,
                 // we'll instead complete the promise at the call site.
                 if (!asyncResult.CompletedSynchronously)
                 {
+                    Debug.Assert(thisRef != null);
```

Comment:
@stephentoub @jkotas is this assumption correct? Or should I just bang the warning away when calling `promise.Complete`?

Reply:
It's correct as long as the IAsyncResult/BeginXx implementation is behaving correctly.  Technically a BeginXx method could be faulty and invoke the callback delegate multiple times, in which case this assert would likely fail.  But since we expect all of our APM implementations using a debug corelib to be correct, we'd want to know about such a problem in our implementation, so the assert seems fine.

Reply:
Great, thanks for confirming and elaborating 😄